pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,628 @@
+	oldNum := c.existingMembers.Len()
+	if add {
+		c.existingMembers.Insert(member)
+		newNum := c.existingMembers.Len()
+		klog.V(2).Infof("Add new member (%s) in local existing memberlist, num %d->%d", member, oldNum, newNum)
+	} else {
+		if !c.existingMembers.Has(member) {
+			return
+		}
+		c.existingMembers.Delete(member)
+		newNum := c.existingMembers.Len()
+		klog.V(2).Infof("Delete member (%s) in local existing memberlist, num %d->%d", member, oldNum, newNum)
+	}
+	if c.nodeName == node.Name {
+		c.updateLocalNodeStatus()
+	}
+}
+
+func (c *Cluster) joinOrLeaveCluster(join bool) {
+	if join {
+		c.memberRWLock.RLock()
+		defer c.memberRWLock.RUnlock()
+		clusterNodes := c.existingMembers.List()
+		n, err := c.mList.Join(clusterNodes)
+		if err != nil {
+			klog.ErrorS(err, "Join cluster failed, cluster nodes %v", clusterNodes)
+			return
+		}
+		klog.V(2).InfoS("Join cluster", "num", n, "cluster nodes", clusterNodes)
+	} else {
+		if err := c.localNodeLeave(); err != nil {
+			klog.ErrorS(err, "Leave cluster failed")
+			return
+		}
+		klog.V(2).InfoS("Leave cluster", "nodeName", c.nodeName)
+	}
+}
+
+func (c *Cluster) localNodeLeave() error {
+	err := c.mList.Leave(defaultLeavePeriod)
+	if err != nil {
+		klog.ErrorS(err, "Local node leave cluster failed")
+		return err
+	}
+	if err := c.mList.Shutdown(); err != nil {
+		klog.ErrorS(err, "Shutdown cluster failed")
+		return err
+	}
+	newML, err := memberlist.Create(c.mlConfig)
+	if err != nil {
+		klog.ErrorS(err, "Create cluster failed")
+		return err
+	}
+	c.mList = newML
+	return nil
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformerHasSynced, c.nodeListerSynced) {
+		return
+	}
+
+	defer func() {
+		if err := c.mList.Shutdown(); err != nil {
+			klog.ErrorS(err, "Shut down cluster failed")
+		}
+		close(c.nodeEventsCh)
+	}()
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case nodeEvent := <-c.nodeEventsCh:
+			c.syncConsistentHashMap(&nodeEvent)
+		case <-c.localNodeEventCh:
+			c.syncLocalNodeStatus()
+		case hashEvent := <-c.conHashEventCh:
+			c.syncConHash(&hashEvent)
+		}
+	}
+}
+
+//func (c *Cluster) localworker(stopCh <-chan struct{}) {
+//	for {
+//		select {
+//		case <-stopCh:
+//			return
+//		case hashEvent := <-c.conHashEventCh:
+//			c.syncConHash(&hashEvent)
+//		}
+//	}
+//}
+
+func (c *Cluster) syncLocalNodeStatus() {
+	shouldJoin := c.shouldJoinCluster(c.nodeName)
+	joined := c.LocalNodeJoined()
+	if joined && !shouldJoin {
+		// filter nodes through nodeSelector from externalIPPools
+		otherClusterNodes := c.syncClusterMembers()
+		c.addExistingMembers(otherClusterNodes...)
+		klog.InfoS("Received a join signal, local node will join cluster", "node name", c.nodeName)
+		c.joinOrLeaveCluster(false)
+	} else if !joined && shouldJoin {
+		klog.InfoS("Received a leave signal, local node will leave cluster", "node name", c.nodeName)
+		c.joinOrLeaveCluster(false)
```

Comment:
I'm confused by this code block. Isn't the first case actually leaving the cluster while the second is joining? The log says opposite things and the paramter of joinOrLeaveCluster is always false.

Anyway, do you still plan to join and leave the memberlist cluster dynamically? Please check my comments in https://github.com/antrea-io/antrea/pull/2186#discussion_r652406324. I think the current code is a little complex to support it and may have bugs like this code block. And I can see some code that doesn't make a lot of sense is for it. For example, if we will use `syncClusterMembers` to get all selected nodes, why we maintain the "existingMembers" via node event? My suggestions of refactoring this part:
1. Remove dynamic joining and leaving cluster in this PR, let's optimize it later.
2. Do not maintain the existingMembers via node events (the current implementation is incorrect as it assumes externalIPPools didn't change when determining a node event cause a node join or leave cluster, for example, a pool first selects this node and changes to not selecting it, then the node's labels changes, the node event handler will think both old and new labels don't match the pool's selector and won't make the node leave the cluster). And just let all event handlers to trigger the dedicated goroutine to determine whether this node should join or leave the cluster and get all selected other nodes.

Reply:
> “I'm confused by this code block. Isn't the first case actually leaving the cluster while the second is joining? The log says opposite things and the paramter of joinOrLeaveCluster is always false.”

mistake，
```
	shouldJoin := c.shouldJoinCluster(c.nodeName)
	joined := c.LocalNodeJoined()
	if !joined && shouldJoin {
		// filter nodes through nodeSelector from externalIPPools
		otherClusterNodes := c.syncClusterMembers()
		c.addExistingMembers(otherClusterNodes...)
		klog.InfoS("Received a join signal, local node will join cluster", "node name", c.nodeName)
		c.joinOrLeaveCluster(true)
	} else if joined && !shouldJoin {
		klog.InfoS("Received a leave signal, local node will leave cluster", "node name", c.nodeName)
		c.joinOrLeaveCluster(false)
	}
```

> "Anyway, do you still plan to join and leave the memberlist cluster dynamically? Please check my comments in #2186 (comment). I think the current code is a little complex to support it and may have bugs like this code block. And I can see some code that doesn't make a lot of sense is for it. For example, if we will use syncClusterMembers to get all selected nodes, why we maintain the "existingMembers" via node event? My suggestions of refactoring this part:
> 
> Remove dynamic joining and leaving cluster in this PR, let's optimize it later.
> Do not maintain the existingMembers via node events (the current implementation is incorrect as it assumes externalIPPools didn't change when determining a node event cause a node join or leave cluster,
>  for example, a pool first selects this node and changes to not selecting it, then the node's labels changes, the node event handler will think both old and new labels don't match the pool's selector and won't make the node leave the cluster). And just let all event handlers to trigger the dedicated goroutine to determine whether this node should join or leave the cluster and get all selected other nodes."

existingMembers can be deleted, as it is used only  when local node join cluster, however the nodes can list and filtered from node informer and externalIPPools.

Remove dynamic joining and leaving cluster in this PR---- it means the node always running in background?

Reply:
Yes, when Egress feature is enabled. Because I believe if a remote Node ever joins the cluster, the local Node doesn't have a way to remove it from its members anyway (I don't find any method Memberlist provides to remove a member, and making the local node leave the cluster and create a new instance is even worse in this case), in which case it will keep trying to connect these left nodes, and if an externalPool selects all Nodes, it has to be this case anyway.